feat: 좋아요(Like) 기능 구현 - CQRS 패턴 적용

hydro00001 opened the pull request:
## 작업 내용

- Product, Store, Seller에 대한 좋아요/좋아요 취소 API 구현
- 좋아요 상태 조회 및 목록 조회 (페이징) API 구현
- Command와 Query 완전 분리 (LikeCommandController / LikeQueryController)
- Domain 중심 설계: 비즈니스 로직을 Domain 레이어에 집중
- Hexagonal Architecture 준수: Port/Adapter 패턴 적용
- Entity 불변성 유지: @Setter 제거, Builder 패턴 사용
- 메서드 3줄 규칙 준수로 가독성 향상

## 특이사항

### 팀 코드 리뷰 반영
- Entity에서 @Setter 제거하여 불변성 보장
- ResponseEntity 제네릭 타입 명시
- Lombok 적절히 활용 (@Getter, @Builder, @NoArgsConstructor 등)
- Validation을 Domain 레이어에서 처리

## API 엔드포인트

### Command
- POST /api/v1/products/{productId}/likes - 상품 좋아요
- DELETE /api/v1/products/{productId}/likes - 상품 좋아요 취소
- POST /api/v1/stores/{storeId}/likes - 매장 좋아요
- DELETE /api/v1/stores/{storeId}/likes - 매장 좋아요 취소
- POST /api/v1/sellers/{sellerId}/likes - 판매자 좋아요
- DELETE /api/v1/sellers/{sellerId}/likes - 판매자 좋아요 취소

### Query
- GET /api/v1/products/{productId}/likes/status - 좋아요 상태 조회
- GET /api/v1/likes/my?type={type}&page={page} - 내 좋아요 목록 조회

## 주요 기술 적용
- CQRS: Command/Query 완전 분리
- Hexagonal Architecture: Domain → Application → API/Infrastructure
- Domain 중심 설계: 비즈니스 로직 Domain에 집중
- Builder 패턴: Entity 생성 시 불변성 보장
- Lombok: 보일러플레이트 코드 최소화 (Domain 제외)

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Implements a unified like system (product/store/seller) with CQRS APIs, domain model, services/ports, JPA adapters, error codes/handlers, and a likes table migration.
> 
> - **API (CQRS)**:
>   - `LikeCommandController`: POST/DELETE for `products/{id}/likes`, `stores/{id}/likes`, `sellers/{id}/likes` returning `ApiResponse<LikeResponse>` or `204`.
>   - `LikeQueryController`: status endpoints (`*/likes/status`) and paged "my likes" (`/users/me/likes/*`) returning `ApiResponse<Page<Object>>` and `LikeStatusResponse`.
>   - `LikeWebMapper`, responses: `LikeResponse`, `LikeStatusResponse`.
> - **Application**:
>   - DTOs: `LikeOutput`, `LikeStatusOutput`, `AddLikeInput`, `DeleteLikeInput`.
>   - Services: `LikeCommandService`, `LikeQueryService` implementing `LikeCommandUseCase`, `LikeQueryUseCase` with duplicate/existence checks.
>   - Ports: inports (`LikeCommandUseCase`, `LikeQueryUseCase`), outports (`LikeCommandPort`, `LikeQueryPort`).
>   - Mapper: `LikeMapper` (domain <-> DTO).
> - **Domain**:
>   - `Like` aggregate with validation and factories; `LikeableType` enum.
> - **Infrastructure (JPA)**:
>   - Adapters: `LikeCommandPersistenceAdapter`, `LikeQueryPersistenceAdapter`.
>   - `LikeEntity`, `LikeEntityMapper`, `LikeRepository` (exists/find/delete/count, paging/sorting).
> - **Error handling & Codes**:
>   - `GlobalExceptionHandler`: handlers for `LikeAlreadyExistsException`, `LikeNotFoundException`, `TargetNotFoundException`.
>   - `ResponseCode`: add `LIKE_*` and `TARGET_NOT_FOUND` entries.
> - **DB**:
>   - Migration `mysql/V1__create_likes_table.sql` creating `likes` with unique and supporting indexes.
> - **Tests**:
>   - Unit tests for services and integration tests for persistence adapters.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 8b5e27640617fa664a1da61b4a6d5bf9c9096456. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

hydro00001 committed:
feat: Likes 도메인 구축 (헥사고날 + DDD + Clean Code)

- 헥사고날 아키텍처 (Ports & Adapters) 적용
- DDD 원칙: Domain Layer 순수 Java (외부 의존성 0)
- Clean Code: 모든 메소드 3줄 이하

[Domain Layer]
- Like.java: 순수 도메인 모델
- LikeableType.java: Enum (PRODUCT, STORE, SELLER)

[Application Layer]
- LikeUseCase: Inbound Port
- LikePort: Outbound Port
- LikeService: 비즈니스 로직 (3줄 이하)
- 예외: LikeAlreadyExistsException, LikeNotFoundException
- 테스트: LikeServiceTest (12개 테스트 케이스)

[Infrastructure Layer]
- LikeEntity: JPA 엔티티
- LikeRepository: Spring Data JPA
- LikePersistenceAdapter: Port 구현체

[API Layer]
- LikeController: 12개 REST API 엔드포인트
  * Product 찜: POST/DELETE/GET (4개)
  * Store 구독: POST/DELETE/GET (4개)
  * Seller 팔로우: POST/DELETE/GET (4개)

[Database]
- likes 테이블 DDL (V1__create_likes_table.sql)
- 다형적 관계 (Polymorphic Association)
- Unique 제약조건 & 인덱스

[Configuration]
- Docker Compose: MySQL 8.0 + Redis 7
- Gradle: Java 21, Kotlin 1.9.25
- 테스트: Mockito, JUnit 5, AssertJ

[Documentation]
- LIKES_SETUP.md: 설정 및 실행 가이드
- API 테스트 예제
- 트러블슈팅 가이드

hydro00001 committed:
docs: 팀원 온보딩 가이드 및 환경 설정 추가

- .env.example 파일 생성 (Docker MySQL 설정 포함)
- README.md에 '처음 시작하기' 섹션 추가
- docker-compose.yml 포트 3306으로 복구
- 팀원이 클론 후 바로 시작할 수 있도록 가이드 작성

hydro00001 committed:
refactor(like): CQRS 패턴 적용 및 Like 도메인 리팩토링

- Command/Query 책임 분리 (Controller, UseCase, Service, Port)
- Lombok과 Builder 패턴 적용으로 보일러플레이트 코드 41% 감소
- 읽기 전용 트랜잭션(@Transactional(readOnly = true)) 최적화
- 메서드 3줄 규칙 적용으로 가독성 향상
- 통합 테스트 추가 (LikePersistenceAdapterIntegrationTest)
- 상세 문서 추가 (CQRS, Lombok, 리팩토링 가이드 등)

Resolves: #N/A

hydro00001 committed:
Merge remote-tracking branch 'origin/develop' into feature/wishList

hydro00001 committed:
chore: 불필요한 문서 파일 제거

- 개발 과정 문서들 제거 (DAILY_REPORT, INTERVIEW_GUIDE 등)
- 핵심 문서만 유지 (README, CQRS_REFACTORING_REPORT, FINAL_STATUS_REPORT)

hydro00001 committed:
refactor(like): Entity에서 @Setter 제거

- 팀 코드 리뷰 반영: Entity는 불변성 유지를 위해 Setter 금지
- Builder 패턴과 생성자만 사용하도록 개선

JiHongKim15 commented on CQRS_REFACTORING_REPORT.md:
불필요한 파일같습니다~

JiHongKim15 replied:
삭제하겠습니다.

JiHongKim15 commented on FINAL_STATUS_REPORT.md:
마찬가지로 md파일 지워주세용

JiHongKim15 replied:
삭제했습니다.

hydro00001 committed:
chore: 불필요한 문서 파일 및 example 파일 제거

JiHongKim15 commented on music-api/src/main/java/com/music/sale/web/like/LikeCommandController.java:
```diff
@@ -0,0 +1,146 @@
+    @DeleteMapping("/products/{productId}/likes")
+    public ResponseEntity<Void> unlikeProduct(
+            @PathVariable Long productId,
+            @RequestHeader("X-User-Id") Long userId
```
user는 나중에 Session에서 꺼내쓰는 형식이 될거라 header에서 받는건 제거해주세요~ controller 안에 String user해서 mock 넣능 느낌으로 하면될것같아요

JiHongKim15 commented on music-api/src/main/java/com/music/sale/web/like/LikeCommandController.java:
```diff
@@ -0,0 +1,146 @@
+     * POST /api/v1/sellers/{sellerId}/likes
+     */
+    @PostMapping("/sellers/{sellerId}/likes")
+    public ResponseEntity<ApiResponse<LikeResponse>> followSeller(
```
APIResponse에 responseEntity가 있지 않나요? 생각해보이 이중 중첩같네요

hydro00001 committed:
revert: 불필요한 설정 파일 수정 되돌리기

- gradle.properties: 개인 로컬 환경 설정 제거
- README.md: 온보딩 가이드 제거
- build.gradle.kts: useJUnitPlatform() 복원
- HELP.md: Spring 자동 생성 파일 삭제

hydro00001 committed:
refactor: 불필요한 주석 제거 및 gradle 버전 복구

- gradle-wrapper.properties: 8.5 -> 8.12.1 복구
- HELP.md 복구
- Copyright 주석 제거
- Controller 메서드 Javadoc 및 섹션 구분 주석 제거

hydro00001 committed:
chore: develop 브랜치 파일 복구

- .env.example 복구
- DOMAIN_LANGUAGES.md 복구
- SETUP.md 복구

hydro00001 committed:
revert: gradle 설정 및 불필요한 파일 변경 제거

- build.gradle.kts 파일들을 develop 버전으로 복구
- start-db.sh 제거 (개인 헬퍼 스크립트)

JiHongKim15 commented on music-application/src/main/java/com/music/sale/application/like/service/LikeCommandService.java:
```diff
@@ -0,0 +1,95 @@
+        return new LikeNotFoundException(getLikeableTypeKorean(type) + " 기록을 찾을 수 없습니다.");
+    }
+
+    private String getLikeableTypeKorean(LikeableType type) {
```
enums로 묶어서 거기서 타입별로 리턴하도록 상수 처리 해주세요. 상수는 반드시 이넘으로 관리 부탁드립니다

hydro00001 replied:
public enum LikeableType {
    PRODUCT("상품 좋아요"),
    STORE("매장 구독"),
    SELLER("판매자 팔로우");
    
    private final String korean;
    
    LikeableType(String korean) {
        this.korean = korean;
    }
    
    public String getKorean() {
        return korean;
    }
}

이렇게 enum으로 묶어서 처리하겠습니다.

JiHongKim15 commented on music-infrastructure/src/main/java/com/music/sale/persistence/like/entity/LikeEntity.java:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * Entity -> Domain 변환
+     */
+    public Like toDomain() {
```
매퍼 만들어서 처리해주세요

hydro00001 replied:
mapper 생성완료했습니다.

hydro00001 committed:
refactor: pr 피드백 반영한 수정

1. LikeableType enum에 korean 필드 추가
   - 상수를 enum으로 관리
   - getLikeableTypeKorean() 메서드 제거

2. Entity의 toDomain() 메서드를 Mapper로 분리
   - LikeEntityMapper 생성
   - Entity ↔ Domain 변환 로직 분리
   - Infrastructure 레이어 책임 명확화

3. User 인증 처리 방식 변경
   - @RequestHeader("X-User-Id") 제거
   - Mock userId (1L) 사용
   - TODO 주석 추가: Session에서 가져오도록 수정 예정

관련 이슈: #6

hydro00001 commented:
피드백 반영 완료했습니다.

PR #9: refactor: pr 내용 반영한 수정
https://github.com/JiHongKim15/music-sale-platform-be/pull/9

재검토 부탁드립니다.

hydro00001 committed:
refactor: 명시적 생성자 제거 및 @RequiredArgsConstructor 적용

JiHongKim15 committed:
Merge pull request #9 from JiHongKim15/refactor/like-feedback

refactor: pr 내용 반영한 수정
